Fix calculation price in group reductions when in tax excluded mode

matthieu-rolland opened the pull request:
| Questions     | Answers
| ------------- | -------------------------------------------------------
| Branch?       | develop
| Description?  | Fix wrong price calculation in grouped buy reduction mode, when taxes are excluded.
| Type?         | bug fix
| Category?     | FO
| BC breaks?    | no
| Deprecations? | no
| Fixed ticket? | Fixes #9929
| How to test?  | In back office, create a product, with a price of 10 euros HT. Then create a special price of 8 euros, starting at 10 products. And another special price of 7 euros, starting at 20 products. Then set the price display to HT (tax excluded). In front office, the product should display correct price reductions.

<!-- Click the form's "Preview" button to make sure the table is functional in GitHub. Thank you! -->

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/prestashop/prestashop/15549)
<!-- Reviewable:end -->

matthieu-rolland committed:
only apply tax rate when calculation method is set to tax-inc

matthieu-rolland committed:
extract price default currency calculation in a dedicated function

matks commented:
@matthieu-rolland do you think it would be possible to add a Behat test to cover this usecase ?

matthieu-rolland commented:
> @matthieu-rolland do you think it would be possible to add a Behat test to cover this usecase ?

Sure, great idea, I'll do that !

Progi1984 commented on controllers/front/ProductController.php:
```diff
@@ -890,7 +890,8 @@ protected function formatQuantityDiscounts($specific_prices, $price, $tax_rate,
 
                 /** @var float $currentPriceDefaultCurrency current price with taxes in default currency */
-                $currentPriceDefaultCurrency = (!$row['reduction_tax'] ? $row['price'] : $row['price'] * (1 + $tax_rate / 100)) + (float) $ecotax_amount;
+                $currentPriceDefaultCurrency = $this->calculatePriceDefaultCurrency($row, $tax_rate, $ecotax_amount);
```
@matthieu-rolland You should send values in parameters and not array. It's not the responsibility of the function to check is the value is in the array.

matthieu-rolland replied:
done !

matthieu-rolland committed:
pass specific values to calculation price function

PierreRambaud commented on controllers/front/ProductController.php:
```diff
@@ -890,7 +890,8 @@ protected function formatQuantityDiscounts($specific_prices, $price, $tax_rate,
 
                 /** @var float $currentPriceDefaultCurrency current price with taxes in default currency */
-                $currentPriceDefaultCurrency = (!$row['reduction_tax'] ? $row['price'] : $row['price'] * (1 + $tax_rate / 100)) + (float) $ecotax_amount;
+                $currentPriceDefaultCurrency = $this->calculatePriceDefaultCurrency($row['price'], $row['reduction_tax'], $tax_rate, $ecotax_amount);
```
```suggestion
                $currentPriceDefaultCurrency = $this->calculatePriceDefaultCurrency(
                    $row['price'], 
                    $row['reduction_tax'], 
                    $tax_rate, $ecotax_amount
                );
```

PierreRambaud commented on controllers/front/ProductController.php:
```diff
@@ -954,6 +955,25 @@ protected function formatQuantityDiscounts($specific_prices, $price, $tax_rate,
+     *
+     * @return mixed
+     */
+    private function calculatePriceDefaultCurrency($specificPriceValue, $specificPriceReductionTax, $taxRate, $ecoTaxAmount)
```
```suggestion
    private function calculatePriceDefaultCurrency(
        $specificPriceValue, 
        $specificPriceReductionTax, 
        $taxRate, 
        $ecoTaxAmount
    ) {
```

PierreRambaud commented on controllers/front/ProductController.php:
```diff
@@ -954,6 +955,25 @@ protected function formatQuantityDiscounts($specific_prices, $price, $tax_rate,
+        $priceDefaultCurrency = $specificPriceValue;
+
+        // only apply tax rate when tax calculation method is set to PS_TAX_INC
+        if (Product::$_taxCalculationMethod == PS_TAX_INC && $specificPriceReductionTax) {
```
```suggestion
        if (Product::$_taxCalculationMethod === PS_TAX_INC && $specificPriceReductionTax) {
```

matthieu-rolland replied:
This is legacy code that I extracted in a function...  (I mean this is how this condition is checked in this controller) I really don't know if Product::$_taxCalculationMethod will always be an integer (and not, let's say, a string).

It felt safer to keep the legacy behavior in this condition. I'll check that out.

PierreRambaud replied:
This value can only be an integer (PS_TAX_EXC and PS_TAX_INC) / null (intiliazed as property):
Feel free to decide :sweat_smile:

matthieu-rolland replied:
Ok, from what I checked in the codebase, it seems that this variable is always assigned an integer. Done.

matthieu-rolland committed:
code style modifications

matthieu-rolland committed:
use stronger condition to check tax display mode

PierreRambaud commented on controllers/front/ProductController.php:
```diff
@@ -954,6 +960,29 @@ protected function formatQuantityDiscounts($specific_prices, $price, $tax_rate,
+    /**
+     * @param float $specificPriceValue
+     * @param bool $specificPriceReductionTax
+     * @param float $ecoTaxAmount
```
`$taxRate` is missing.

matthieu-rolland replied:
yep, done

PierreRambaud commented on controllers/front/ProductController.php:
```diff
@@ -954,6 +960,29 @@ protected function formatQuantityDiscounts($specific_prices, $price, $tax_rate,
+     *
+     * @return mixed
+     */
+    private function calculatePriceDefaultCurrency(
```
```suggestion
    private function calculatePriceDefaultCurrencyAmount(
```

You're not calculating a currency. :thinking: 
Wdyt @matks?

matthieu-rolland replied:
It only calculates a price, in default currency because we take data from a specific price without applying another currency.

I gave this name according to the variable it fills in the parent function, but you are right, this function has no knowledge of any currency matters. I think I should change that. Let me think of a better name.

matthieu-rolland replied:
I called it calculateSpecificPrice, because in the end this is what it really does, for a specific price basis, it calculates its final value (with or without tax rate and ecotax)

matthieu-rolland committed:
add missing phpdoc param

matthieu-rolland committed:
rename calculation price method

PierreRambaud commented on controllers/front/ProductController.php:
```diff
@@ -954,6 +960,30 @@ protected function formatQuantityDiscounts($specific_prices, $price, $tax_rate,
+        $taxRate,
+        $ecoTaxAmount
+    ) {
+        $priceDefaultCurrency = $specificPriceValue;
```
You forget to change the variable name :trollface:

matthieu-rolland replied:
👍 :trollface:

matthieu-rolland committed:
Extract formatQuantityDiscounts method in a proper class to ease unit testing

matthieu-rolland committed:
remove productcontrollertest and put the (improved) tests in SpecificPriceFormatterTest

matthieu-rolland committed:
Add class comment

matthieu-rolland committed:
add comment

matthieu-rolland commented on src/Core/SpecificPrice/SpecificPriceFormatter.php:
```diff
@@ -0,0 +1,182 @@
+<?php
```
This class is an extraction of the legacy function **formatQuantityDiscounts**, it is still legacy code, I extracted it so I could pass it some values has dependency, especially the **isTaxIncluded** boolean, so it can be tested with and without tax.

matthieu-rolland commented on tests-legacy/Unit/Core/SpecificPrice/SpecificPriceFormatterTest.php:
```diff
@@ -24,39 +24,29 @@
  * International Registered Trademark & Property of PrestaShop SA
  */
```
This is a legacy test that I moved in a new test file (since the code tested is in another class).
I just added some testing without taxes included.

matthieu-rolland committed:
cs fixer

matthieu-rolland commented:
> @matthieu-rolland do you think it would be possible to add a Behat test to cover this usecase ?

So I updated the unit test of the method calculating the discount value. It now both tests tax included and tax ecluded.

About the behat test, I started doing it, but I realised it was testing only this function, which is exactly what the unit test is already doing.
A proper behat test would imply starting from the product, its specific prices, testing the currencies, the language locale, the context... and many other things. Which I think is way out of the scope of this PR.

Progi1984 commented on src/Core/SpecificPrice/SpecificPriceFormatter.php:
```diff
@@ -0,0 +1,182 @@
+     * @param bool $isTaxIncluded
+     * @param Context $context
+     */
+    public function __construct(array $specificPrice, bool $isTaxIncluded, $context)
```
@matthieu-rolland It lacks
```suggestion
    public function __construct(array $specificPrice, bool $isTaxIncluded, Context $context)
```

jolelievre replied:
@matthieu-rolland Indeed just this modification left and we're good to go ;)

matthieu-rolland replied:
done !

Progi1984 commented on src/Core/SpecificPrice/SpecificPriceFormatter.php:
```diff
@@ -0,0 +1,182 @@
+     *
+     * @return array
+     */
+    public function formatSpecificPrice($initialPrice, $tax_rate, $ecotax_amount)
```
@matthieu-rolland Isn't it possible to divide this complex method into small private methods ?

matthieu-rolland replied:
@Progi1984 
As I wrote earlier, this is a legacy method that I put in a dedicated class in order to be able to inject the "taxincluded" boolean as a dependancy, making it possible to fix the unit test too.

The goal of this PR is not to refactor this part, I tried though, but it's too complex and way out of the scope of this PR. I only refactored the small part that was causing the bug.

jolelievre commented on tests-legacy/Unit/Core/SpecificPrice/SpecificPriceFormatterTest.php:
```diff
@@ -0,0 +1,244 @@
+        $specificPriceFormatter = new SpecificPriceFormatter($specificPrices[0], $isTaxIncluded, $context);
+        $formattedSpecificPrice = $specificPriceFormatter->formatSpecificPrice($price, $taxRate, $ecotaxAmount);
+
+        $priceFormatter = new PriceFormatter();
```
I'm not sure I understand this test You actually compare both behaviours and test if they match?
But then it's not really a unit test

matthieu-rolland replied:
@jolelievre  There is only one method tested, it's a legacy method that returns an array, I'm checking that the **discount** and **save** values are each matching their respective expectations coming from the data provider.

Behaviors are specified in the provider (if we can call that a behavior). This is a legacy test that I had to move because I moved the method that is tested, the logic has not changed, I just added cases in the data provider, and changed the way the method is called, because it has moved in its own class.

matthieu-rolland commented on src/Core/SpecificPrice/SpecificPriceFormatter.php:
```diff
@@ -0,0 +1,182 @@
+        if ($this->isTaxIncluded && $specificPriceReductionTax) {
+            $specificPrice = $specificPrice * (1 + $taxRate / 100) + (float) $ecoTaxAmount;
+        }
+
```
This is the actual fix

matthieu-rolland committed:
add context type in SpecificPriceFormatter constructor signature

PierreRambaud commented on src/Core/SpecificPrice/SpecificPriceFormatter.php:
```diff
@@ -0,0 +1,182 @@
+     * @param bool $isTaxIncluded
+     * @param Context $context
+     */
+    public function __construct(array $specificPrice, bool $isTaxIncluded, \Context $context)
```
Use namespace instead, better for refactoring :)

```suggestion
    public function __construct(array $specificPrice, bool $isTaxIncluded, Context $context)
```

matthieu-rolland replied:
@PierreRambaud  Context has no namespace... but:

Your comment makes me realize that this class, that contains legacy code, in which I only refactored the part where my fix is applied, should not be in core, but in the legacy code. I'll move that (and change the calls accordingly)

PierreRambaud replied:
When I said, use namespace, I mean `use Context;` at the top to import them as global classes.

matthieu-rolland replied:
done !

PierreRambaud commented on src/Core/SpecificPrice/SpecificPriceFormatter.php:
```diff
@@ -0,0 +1,182 @@
+            $this->specificPrice['real_value'] = $initialPrice > 0 ? $initialPrice - $currentPriceCurrentCurrency : $currentPriceCurrentCurrency;
+            $discountPrice = $initialPrice - $this->specificPrice['real_value'];
+
+            if (\Configuration::get('PS_DISPLAY_DISCOUNT_PRICE')) {
```
Same for Configuration

matthieu-rolland replied:
done !

matks commented:
@matthieu-rolland  it seems you wait for @PierreRambaud  validation of the changes you added following his code review, is that correct ?

matthieu-rolland commented:
> @matthieu-rolland it seems you wait for @PierreRambaud validation of the changes you added following his code review, is that correct ?

Not really, after talking after other devs from the core team, we concluded that I need @eternoendless 's opinion about this matter:

I extracted a complex legacy function in a class (for easy unit testing purpose), that I put in the core. But the class still contains spaghetti legacy code (only the part I fixed is refactored). 

I was not sure if this class should go in the legacy or stay in the core (knowing it has legacy in it) ?

@eternoendless your opinion would be appreciated 🤔

eternoendless commented on src/Core/SpecificPrice/SpecificPriceFormatter.php:
```diff
@@ -0,0 +1,184 @@
+namespace PrestaShop\PrestaShop\Core\SpecificPrice;
+
+use Configuration;
+use Context;
```
Core code SHOULD NOT depend on legacy classes. If you have legacy dependencies, then you MUST declare an interface and implement it as an Adapter. Also, try to keep legacy dependencies out of the interface.

matthieu-rolland replied:
Thank you for your input @eternoendless I'll do that 👍

matthieu-rolland replied:
So as we said, there is no adapter, a legacy class was created (it will never be used by the core, and no new code was added, it's just a legacy method being moved in a new legacy class due to unit testing issues).

matthieu-rolland committed:
move class to legacy, change test class and calls accordingly

jolelievre commented on classes/product/SpecificPriceFormatter.php:
```diff
@@ -0,0 +1,179 @@
+            $this->specificPrice['real_value'] = $initialPrice > 0 ? $initialPrice - $currentPriceCurrentCurrency : $currentPriceCurrentCurrency;
+            $discountPrice = $initialPrice - $this->specificPrice['real_value'];
+
+            if (Configuration::get('PS_DISPLAY_DISCOUNT_PRICE')) {
```
Maybe you can inject this configuration in the constructor as well?
So that we limit as much as possible the references to other legacy classes

matthieu-rolland replied:
@jolelievre  done !

jolelievre commented on classes/product/SpecificPriceFormatter.php:
```diff
@@ -0,0 +1,179 @@
+            // Since this price is set in default currency,
+            // we need to convert it into current currency
+            $this->specificPrice['id_currency'];
+            $currentPriceCurrentCurrency = \Tools::convertPrice($currentPriceDefaultCurrency, $this->context->currency, true, $this->context);
```
Seems like you only need the context to use this method
But if you look at the `Tools::convertPrice` it only needs the context as a fallback if the currency parameter is not provided
So instead of injecting the context you could simply inject the currency or use it as a parameter of your method, this way you could call `\Tools::convertPrice($currentPriceDefaultCurrency, $this->currency, true);` with the same result but then you can remove the dependency to Context

matthieu-rolland replied:
@jolelievre  done !

matthieu-rolland committed:
minor refactor

jolelievre reviewed (approved):
Nice work @matthieu-rolland

PierreRambaud commented:
Thanks @matthieu-rolland
